Add some new variables

Gamer153 opened the pull request:
`%PREVIOUSPAGE%` or `%PREVPAGE%` - There is currently a `%NEXTPAGE%` variable which can grab the next page of history (assuming it does currentpage + 1) but there was no previous page variable.

`%TIME_PASSED_SIMPLE%`: Currently, the `%TIME_PASSED%` variable displays very specific (example: 6 months, 3 weeks, 6 days, 14 hours, 12 seconds) so this is just a simpler version (example: 7 months. Rounded up from 6 months, 3 weeks, 6 days)

`%TIME_REMAINING_SIMPLE%`: The same as `%TIME_PASSED_SIMPLE%`, but for `%TIME_REMAINING%`

`%TYPE_VERB%` - Currently `%TYPE%` exists which outputs Ban, Warn, Mute, Kick, but this variable outputs Banned, Warned, Muted, Kicked instead. (configurable)

`%IS_ACTIVE%` Is this punishment active? Check if ban/mute has remaining time, or if warn or kick has been revoked. Returns a configurable output depending on if the punishment is active or expired.

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/Formatter.java:
```diff
@@ -215,12 +229,16 @@ String getVariable() {
 		simpleReplacements.put(SimpleReplaceable.TIME_REMAINING, relativeEndFormatted);
+		simpleReplacements.put(SimpleReplaceable.TIME_REMAINING_SIMPLE, relativeEndFormattedSimple);
+		MessagesConfig.Formatting.PunishmentActiveDisplay display = messages().formatting().punishmentActiveDisplay();
+		simpleReplacements.put(SimpleReplaceable.IS_ACTIVE, (display != null) ? ((isActive) ? display.active() : display.expired()) : "Unknown");
```
The null check isn't necessary, because the configuration will update itself by adding default configuration options if they do not exist.
```suggestion
		simpleReplacements.put(SimpleReplaceable.IS_ACTIVE, (isActive) ? display.active() : display.expired());
```

Gamer153 replied:
That didn't work for me though, the tests kept failing because `punishmentActiveDisplay()` returned null. It worked flawlessly ingame.

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/Formatter.java:
```diff
@@ -355,13 +373,45 @@ String formatRelative(long diff) {
 		return joiner.toString();
 	}
+
+	String formatRelativeSimple(long diff) {
```
Ideally there should be a unit test for this. The current tests will ensure that it does not throw an exception, but for completeness' sake, we should verify the output as well.

While a small unit test would be ideal I won't consider it a barrier to merging. It doesn't have to be a large test; also, FormatterTest should have most of the setup done for you.

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/MessagesConfig.java:
```diff
@@ -232,6 +232,33 @@ default Component prefix() {
+
+		interface PunishmentActiveDisplay {
+
+			@ConfComments("How do you describe an active punishments stats?")
```
```suggestion
			@ConfComments("How do you describe an active punishment?")
```

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/MessagesConfig.java:
```diff
@@ -232,6 +232,33 @@ default Component prefix() {
+			@DefaultString("Active")
+			String active();
+
+			@ConfComments("How do you describe an expired punishments state?")
```
```suggestion
			@ConfComments("How do you describe an expired punishment?")
```

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/MessagesConfig.java:
```diff
@@ -232,6 +232,33 @@ default Component prefix() {
 		Map<PunishmentType, String> punishmentTypeDisplay();
 
+		@ConfKey("punishment-type-verb-display")
+		@ConfComments("How should punishment types be displayed as a verb?")
```
```suggestion
		@ConfComments("How should punishment types be displayed as a verb? Used for the %TYPE_VERB% variable")
```

A248 reviewed (commented):
Looks very good! The work you have done is much appreciated.

You are just in time for the release of 1.0.0, if we can manage to merge this soon. I'm very glad to receive a PR at this time on holiday.

One more note about the code: there is a place in MessagesConfig where the variables are documented in a comment. The new variables should be added there: https://github.com/A248/LibertyBans/blob/709bbae14f0d27fd7a86237139f78d9a44d30ac8/bans-core/src/main/java/space/arim/libertybans/core/config/MessagesConfig.java#L52-L55

Feel free to also add yourself to [the contributors file](https://github.com/A248/LibertyBans/blob/master/bans-core/src/main/resources/contributors). The format is maintainers, collaborators, and commended users -- as you can see in `/libertybans about`.

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/Formatter.java:
```diff
@@ -172,40 +176,50 @@ String getVariable() {
 
-		final String durationFormatted;
-		final String relativeEndFormatted;
+		final String durationFormatted, durationFormattedSimple;
```
Is there a reason you decided to put the `durationFormattedSimple` variable here? As far as I can see, it's only used in one place. You don't even need the variable at all.

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/commands/ListCommands.java:
```diff
@@ -216,11 +216,14 @@ private void noPunishmentsOnThisPage(int page) {
-							.replace("%NEXTPAGE%", nextPageString);
+							.replace("%NEXTPAGE%", nextPageString)
+							.replace("%PREVIOUSPAGE%", previousPageString)
+							.replace("%PREVPAGE%", previousPageString);
```
There should be only one previous page variable. I suggest `%PREVIOUSPAGE%`

A248 commented:
I realized the `%IS_ACTIVE%` variable as implemented in this PR will not be accurate. Checking whether the punishment time has elapsed is insufficient to determine whether the punishment is active -- the punishment can be revoked before it exppires, and thereby no longer be active.

I suggest removing the `%IS_ACTIVE%` variable. It can't be implemented well in the present state of affairs. The necessary contextual information is not there.

Gamer153 commented:
- Added a unit test for `formatRelativeSimple()`.
- Added the variables to the docs. (`%NEXTPAGE%` and `%PREVPAGE%` too.)
- Added myself to the contributers file.
- Renamed `%IS_ACTIVE%` to `%HAS_EXPIRED%` so that it now reflects what it actually stands for.
- Removed unnecessary variable.
- Removed `%PREVPAGE%`.
- Improved docs.

This is a short summary of my changes and I should've implemented all the changes you wanted.

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/InternalFormatter.java:
```diff
@@ -21,6 +21,7 @@
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.ComponentLike;
 import space.arim.api.jsonchat.adventure.util.ComponentText;
+import space.arim.libertybans.api.PunishmentType;
```
Remove this stray import

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/MessagesConfig.java:
```diff
@@ -232,6 +238,33 @@ default Component prefix() {
+		@ConfKey("punishment-expired-display")
+		@ConfComments("How should the %HAS_EXPIRED% variable be displayed?")
+		@SubSection
+		PunishmentExpiredDisplay punishmentActiveDisplay();
```
```suggestion
		PunishmentExpiredDisplay punishmentExpiredDisplay();
```
To be consistent with the rest of the code

A248 commented on bans-core/src/main/java/space/arim/libertybans/core/config/Formatter.java:
```diff
@@ -215,12 +228,16 @@ String getVariable() {
 		simpleReplacements.put(SimpleReplaceable.TIME_REMAINING, relativeEndFormatted);
+		simpleReplacements.put(SimpleReplaceable.TIME_REMAINING_SIMPLE, relativeEndFormattedSimple);
+		MessagesConfig.Formatting.PunishmentExpiredDisplay display = messages().formatting().punishmentActiveDisplay();
+		simpleReplacements.put(SimpleReplaceable.HAS_EXPIRED, (display != null) ? ((notExpired) ? display.notExpired() : display.expired()) : "Unknown"); // null check needed for tests
```
A better way to make the tests work is to mock the PunishmentExpiredDisplay in the test setup, rather than add a null check to the main code.

If you modify the `setSimpleMessagesFormatting` method in `FormatterTest`, you can do this without much trouble:
```java
	private void setSimpleMessagesFormatting() {
  // These lines already exist
		MessagesConfig.Formatting formatting = mock(MessagesConfig.Formatting.class);
		when(messagesConfig.formatting()).thenReturn(formatting);
		lenient().when(formatting.consoleDisplay()).thenReturn("Console");
		lenient().when(formatting.globalScopeDisplay()).thenReturn("global");
		lenient().when(formatting.punishmentTypeDisplay()).thenReturn(Map.of());
		lenient().when(formatting.noTimeRemainingDisplay()).thenReturn("(No time remaining)");
  // Adding these lines will allow you to remove the null check
		MessagesConfig.Formatting.PunishmentExpiredDisplay expiredDisplay = mock(MessagesConfig.Formatting.PunishmentExpiredDisplay.class);
                lenient().when(expiredDisplay.notExpired()).thenReturn("not expired");
                lenient().when(expiredDisplay.expired()).thenReturn("expired");
                lenient().when(formatting.punishmentExpiredDisplay()).thenReturn(expiredDisplay);
	}
```

A248 reviewed (commented):
Almost done. The important item left is to remove the null check by modifying the tests appropriately.
